graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -48,6 +53,283 @@ std::pair<torch::Tensor, torch::Tensor> Sort(
+      const auto u = indices[in_idx];
+      hop[out_row + rofs] = u;
+    }
+    tx += stride_x;
+  }
+}
+
+// Given rows and indptr, computes:
+// inrow_indptr[i] = indptr[rows[i]];
+// in_degree[i] = indptr[rows[i] + 1] - indptr[rows[i]];
+template <typename indptr_t, typename nodes_t>
+struct DegreeFunc {
+  const nodes_t* rows;
+  const indptr_t* indptr;
+  indptr_t* in_degree;
+  indptr_t* inrow_indptr;
+  __host__ __device__ auto operator()(int64_t tIdx) {
+    const auto out_row = rows[tIdx];
+    const auto indptr_val = indptr[out_row];
+    const auto degree = indptr[out_row + 1] - indptr_val;
+    in_degree[tIdx] = degree;
+    inrow_indptr[tIdx] = indptr_val;
+  }
+};
+
+struct PairSum {
+  template <typename indptr_t>
+  __host__ __device__ auto operator()(
+      thrust::tuple<indptr_t, indptr_t> a,
+      thrust::tuple<indptr_t, indptr_t> b) {
+    return thrust::make_tuple(
+        thrust::get<0>(a) + thrust::get<0>(b),
+        thrust::get<1>(a) + thrust::get<1>(b));
+  };
+};
+
+template <typename indptr_t>
+auto ComputeDegree(
+    const indptr_t* const indptr, torch::Tensor nodes, cudaStream_t stream) {
+  auto allocator = cuda::BuildAllocator();
+  const auto exec_policy = thrust::cuda::par_nosync(allocator).on(stream);
+  const int64_t num_nodes = nodes.size(0);
+  // Read indptr only once in case it is pinned and access is slow.
+  auto sliced_indptr = allocator.AllocateStorage<indptr_t>(num_nodes);
+  // compute in-degrees
+  auto in_degree = allocator.AllocateStorage<indptr_t>(num_nodes + 1);
+  thrust::counting_iterator<int64_t> iota(0);
+  AT_DISPATCH_INDEX_TYPES(nodes.scalar_type(), "IndexSelectCSCNodes", ([&] {
+                            using nodes_t = index_t;
+                            thrust::for_each(
+                                exec_policy, iota, iota + num_nodes,
+                                DegreeFunc<indptr_t, nodes_t>{
+                                    nodes.data_ptr<nodes_t>(), indptr,
+                                    in_degree.get(), sliced_indptr.get()});
+                          }));
+  return std::make_pair(std::move(in_degree), std::move(sliced_indptr));
+}
+
+template <typename indptr_t, typename indices_t>
+std::tuple<torch::Tensor, torch::Tensor> UVAIndexSelectCSCCopyIndices(
+    torch::Tensor indices, const indptr_t* const sliced_indptr,
+    const int64_t num_nodes, const indptr_t* const in_degree,
+    const int64_t* const perm, torch::TensorOptions nodes_options,
+    torch::ScalarType indptr_scalar_type, cudaStream_t stream) {
+  auto allocator = cuda::BuildAllocator();
+  thrust::counting_iterator<int64_t> iota(0);
+
+  // Output indptr for the slice indexed by nodes.
+  auto sub_indptr =
+      torch::empty(num_nodes + 1, nodes_options.dtype(indptr_scalar_type));
+
+  // Actual and modified number of edges.
+  indptr_t hop_size, hop_size_aligned;
+  auto sub_indptr_aligned = allocator.AllocateStorage<indptr_t>(num_nodes + 1);
+  {
```

Comment:
why we need {}?

Reply:
Because I didn't want to crowd the main function with temporary variables with limited lifetimes, guarding such regions like that communicates to the person reading the code that these variables and functions are grouped and are used only inside the region.